File: Rust/binary-trees/binarytrees.rs
// The Computer Language Benchmarks Game
// http://benchmarksgame.alioth.debian.org/
//
// contributed by the Rust Project Developers
// contributed by Matt Brubeck
// contributed by TeXitoi
// contributed by Cristi Cobzarenco

extern crate typed_arena;
extern crate rayon;

use typed_arena::Arena;
use rayon::prelude::*;

struct Tree<'a> {
    children: Option<(&'a Tree<'a>, &'a Tree<'a>)>,
}

fn item_check(tree: &Tree) -> i32 {
    if let Some((left, right)) = tree.children {
        1 + item_check(right) + item_check(left)
    } else {
        1
    }
}

fn bottom_up_tree<'r>(arena: &'r Arena<Tree<'r>>, depth: i32)
                  -> &'r Tree<'r> {
    let mut tree = arena.alloc(Tree { children: None });
    if depth > 0 {
        let right = bottom_up_tree(arena, depth - 1);
        let left = bottom_up_tree(arena, depth - 1);
        tree.children = Some((left, right))
    }
    tree
}

fn inner(depth: i32, iterations: i32) -> String {
    let chk: i32 = (0 .. iterations).into_par_iter().map(|_| {
        let arena = Arena::new();
        let a = bottom_up_tree(&arena, depth);
        item_check(a)
    }).sum();
    format!("{}\t trees of depth {}\t check: {}", iterations, depth, chk)
}

fn main() {
    let n = std::env::args().nth(1)
        .and_then(|n| n.parse().ok())
        .unwrap_or(10);
    let min_depth = 4;
    let max_depth = if min_depth + 2 > n { min_depth + 2 } else { n };

    {
        let arena = Arena::new();
        let depth = max_depth + 1;
        let tree = bottom_up_tree(&arena, depth);
        println!("stretch tree of depth {}\t check: {}", depth, item_check(tree));
    }

    let long_lived_arena = Arena::new();
    let long_lived_tree = bottom_up_tree(&long_lived_arena, max_depth);

    let messages = (min_depth/2..max_depth/2 + 1).into_par_iter().map(|half_depth| {
            let depth = half_depth * 2;
            let iterations = 1 << ((max_depth - depth + min_depth) as u32);
            inner(depth, iterations)
        }).collect::<Vec<_>>();

    for message in messages {
        println!("{}", message);
    }

    println!("long lived tree of depth {}\t check: {}", max_depth, item_check(long_lived_tree));
}

File: PHP/binary-trees/binarytrees.php
<?php
/* The Computer Language Benchmarks Game
   http://benchmarksgame.alioth.debian.org/

   contributed by Yuriy Moskalchuk
   (based on contributions of Peter Baltruschat and Levi Cameron)
*/

function createTree($depth)
{
    global $obj;
    if ($depth === 0) {
        return clone $obj;
    }
    $depth--;

    $t = clone $obj;
    $t->l = createTree($depth);
    $t->r = createTree($depth);

    return $t;
}

function checkTree(&$treeNode)
{
    $sum = 1
        + ($treeNode->l->l === null ? 1 : checkTree($treeNode->l))
        + ($treeNode->r->l === null ? 1 : checkTree($treeNode->r));
    unset($treeNode->l);
    unset($treeNode->r);

    return $sum;
}
$obj = new class {
    public $l;
    public $r;
};
$minDepth = 4;
$n = ($argc == 2) ? $argv[1] : 1;
$maxDepth = max($minDepth + 2, $n);
$stretchDepth = $maxDepth + 1;

$stretchTree = createTree($stretchDepth);
echo "stretch tree of depth $stretchDepth\t check: ", checkTree($stretchTree), PHP_EOL;
unset($stretchTree);

$longLivedTree = createTree($maxDepth);

$workersAvailable = file_exists('/proc/cpuinfo') ?
    preg_match_all('/^processor\s/m', file_get_contents('/proc/cpuinfo'), $null) - 1 : 1;
$workersAvailable = $workersAvailable ?: 1;

function startWorker($minDepth, $iterations, &$PIDs, $n, $size, $shmId)
{
    $pid = pcntl_fork();
    if ($pid) {
        $PIDs[$minDepth] = $pid;
    } else {
        $check = 0;
        for ($i = 1; $i <= $iterations; ++$i) {
            $t = createTree($minDepth);
            $check += checkTree($t);
            unset($t);
        }
        $result = "$iterations\t trees of depth $minDepth\t check: $check\n";
        shmop_write($shmId, $result, $size * ($n - 1));
        exit(0);
    }
}

function waitWorkers(&$PIDs, &$workersAvailable)
{
    foreach ($PIDs as $PID) {
        $pid = pcntl_waitpid($PID, $status);
        if ($pid) {
            unset($PIDs[array_search($pid, $PIDs)]);
            $workersAvailable++;
        }
    }
}

$iterations = 1 << ($maxDepth);
$depthIterations = [$minDepth => $iterations];
do {
    $minDepth += 2;
    if ($minDepth <= $maxDepth) {
        $iterations >>= 2;
        $depthIterations[$minDepth] = $iterations;
    }
} while ($minDepth <= $maxDepth);

$size = 128;
extension_loaded('shmop') or dl('shmop.so');
$shmId = shmop_open(ftok(__FILE__, 'b'), 'c', 0644, $size * (count($depthIterations) + 1));
$workersPIDs = [];
$n = 0;
foreach ($depthIterations as $depth => $iteration) {
    if ($workersAvailable > 0) {
        work:
        startWorker($depth, $iteration, $workersPIDs, ++$n, $size, $shmId);
        $workersAvailable--;
    } else {
        waitWorkers($workersPIDs, $workersAvailable);
        if ($workersAvailable) {
            goto work;
        }
    }
}
waitWorkers($workersPIDs, $workersAvailable);

foreach (range(1, count($depthIterations)) as $n) {
    echo trim(shmop_read($shmId, $size * ($n - 1), $size)), PHP_EOL;
}
shmop_delete($shmId);

echo "long lived tree of depth $maxDepth\t check: ", checkTree($longLivedTree), PHP_EOL;